pennylane/devices/qutrit_mixed/apply_operation.py

```diff
@@ -0,0 +1,201 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to apply operations to a qutrit mixed state."""
+# pylint: disable=unused-argument
+
+from functools import singledispatch
+from string import ascii_letters as alphabet
+import pennylane as qml
+from pennylane import math
+from pennylane import numpy as np
+from pennylane.operation import Channel
+from .utils import QUDIT_DIM, get_einsum_mapping, get_new_state_einsum_indices
+
+
+alphabet_array = np.array(list(alphabet))
+
+
+def apply_operation_einsum(op: qml.operation.Operator, state, is_state_batched: bool = False):
+    r"""Apply a quantum channel specified by a list of Kraus operators to subsystems of the
+    quantum state. For a unitary gate, there is a single Kraus operator.
+
+    Args:
+        op (Operator): Operator to apply to the quantum state
+        state (array[complex]): Input quantum state
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        array[complex]: output_state
+    """
+
+    def map_indices(**kwargs):
```

Comment:
I would make this a private function `_map_einsum_indices` or something similar outside `apply_operation_einsum` in the same file.

Reply:
The reason this is in a separate file is to reduce copied code in the next PR. It calls this function as well because it needs to find A@rho. I could add a function to apply_op that does this, but it doesn't really fit.

Reply:
Sorry for the confusion. I'm referring to the `map_indices` function defined inside `apply_operation_einsum()`, not one of the helpers in the utils file.

Reply:
Sorry for the confusion. I'm referring to the map_indices function defined inside apply_operation_einsum(), not one of the helpers in the utils file.

Reply:
Sorry for the confusion. I'm referring to the map_indices function defined inside apply_operation_einsum(), not one of the helpers in the utils file.

Reply:
that makes much mores sense, thank you for clarifying I agree.

Reply:
I also thought this was the utils function, I missread where it was thank you.